### ISSUE TITLE
how to change the weights in nngraph?

### ISSUE THREAD

**Issue by brisker** [nhynes/caffegraph#6]
how to change the weights in nngraph?
What I want to do is--

model = nn.gModule({},{})
para = model:parameters()

para[i] = Torch.Tensor()  --change para[1] accordingly

**Comment by nhynes** [comment]
Unless you're confident that you know exactly which entry of `parameters` corresponds to your module, I recommend finding it by iterating through all modules using one of `Module:(apply|listModules|findModules)` and then changing the parameters on the module, directly.

Note that you should not reassign the Tensor, otherwise it will not point to the Storage that you just created using `Module:parameters()`.

**Comment by brisker** [comment]
@nhynes
Thanks for your reply.
Besides，I want to ask that if I have a nngraph network with multi outputs，like-
……
out2= nn.conv()(out1)  
model=nn.gModule(input,{out1,out2} ) 
df_do1=criterion:backward(input,out1) 
df_do2=criterion:backward(input,out2) 
and in the back prop process, all I need to do is to run the code below even if  there is a connection between out1 and out2, right?


model:backward(input,{df_do1, df_do2}

**Comment by nhynes** [comment]
[Yes, that's correct.](https://github.com/torch/nngraph#a-network-with-2-inputs-and-2-outputs)

**Comment by brisker** [comment]
@nhynes 
Hi, I was trying reproduce the [HED Network](https://github.com/s9xie/hed/blob/master/examples/hed/train_val.prototxt) using Torch, and this network has 1 input and 6 outputs. The original caffe implementation has several hype-parameters for convolution operations, like learning rate and weight decay. I follow all the parameters used in [HED Network](https://github.com/s9xie/hed/blob/master/examples/hed/train_val.prototxt), but still the result seems to be wrong. Could you please give some advice? Thanks a lot!
The model I built --
[generate_model.txt](https://github.com/nhynes/caffegraph/files/852350/generate_model.txt)
The overall training loop of mine is like:
        outputs = model:forward(input) 

	 f1 = criterion:forward(outputs[1], tensorGt)
	 f2 = criterion:forward(outputs[2], tensorGt)
	f3 = criterion:forward(outputs[3], tensorGt)
	 f4 = criterion:forward(outputs[4], tensorGt)
	 f5 = criterion:forward(outputs[5], tensorGt)
	 f_fuse = criterion:forward(outputs[6], tensorGt)

        df_do1 = criterion:backward(outputs[1], tensorGt)
       df_do2 = criterion:backward(outputs[2], tensorGt)
        df_do3 = criterion:backward(outputs[3], tensorGt)
         df_do4 = criterion:backward(outputs[4], tensorGt)
         df_do5 = criterion:backward(outputs[5], tensorGt)
         df_do_fuse = criterion:backward(outputs[6], tensorGt)	
	
	model:backward(input,{df_do1,df_do2,df_do3,df_do4,df_do5,df_do_fuse})

**Comment by nhynes** [comment]
Okay, well there are a few problems here. The first and most important of which is that the HED network uses Deconvolutions, which aren't yet supported in caffegraph. You can add these by looking at the code that does the regular convolutions.

Next is that you're using the same criterion for every backward. If you look at the [code](https://github.com/torch/nn/blob/master/BCECriterion.lua) you'll find that the `gradInput` you're assigning to each df_do* is actually the _exact same tensor_, so you're actually backpropagating the same thing to every branch. Try this
```
local criterion = nn.ParallelCriterion()
for i=1,6 do
  criterion:add(nn.BCECriterion())
end

model:forward(input)
criterion:forward(model.output, targets) -- targets is 6*whatever
model:backward(input, targets)
```

**Comment by brisker** [comment]
terribly sorry that I have deleted my post **by mistake**, but I have already tried deconv, and the model does not converge, seems that the problem is about the criterion, I will give it  a try, thanks so much !
@nhynes

**Comment by brisker** [comment]
@nhynes 
model:backward(input, targets)
or
model:backward(input, gradients)??

**Comment by nhynes** [comment]
`model:backward(input, criterion.gradInput)`, sorry

**Comment by brisker** [comment]
@nhynes 
What if I want to print the 6 losses?

     f = criterion:forward(outputs,{tensorGt,tensorGt,tensorGt,tensorGt,tensorGt,tensorGt})

this line gives a f which is just a 1*1 tensor

**Comment by nhynes** [comment]
```lua
for i,crit in pairs(criterion.criterions) do
  print("Loss["..i.."] = "..crit.output)
end
```

**Comment by brisker** [comment]
@nhynes 
like this: ??


        outputs = model:forward(inputAux)
	f = criterion:forward(outputs,{tensorGt,tensorGt,tensorGt,tensorGt,tensorGt,tensorGt})
	for i,crit in pairs(criterion.criterions) do
  		print("Loss["..i.."] = "..crit.output)
	end
	df_do = criterion:backward(outputs,{tensorGt,tensorGt,tensorGt,tensorGt,tensorGt,tensorGt})
	
	model:backward(inputAux,df_do)

**Comment by nhynes** [comment]
Actually, since you're backwarding the same thing, you can do
```lua
criterion = nn.ParallelCriterion(true) -- repeat the target
-- ...
criterion:forward(outputs, tensorGt)
-- ...
criterion:backward(outputs, tensorGt)
```

Otherwise that looks reasonable! It the model still won't converge, it's probably because you're missing the Deconv (unless you've correspondingly modified the ground truth) or the data loader is loading the wrong thing.

**Comment by brisker** [comment]
@nhynes 
If I use deconv, after 3 epoches, the outputs of the model looks like this:
![image](https://cloud.githubusercontent.com/assets/13804492/24073007/0ee5454c-0c2b-11e7-9ae5-13ffdafa9f8d.png)
clearly wrong(I have tested the original caffe HED code, and after 3 epoches, there are several edges on the output maps, not like this, definitely. **And using the original caffe implementation, my dataset converges well)**
and if I use upsamplingBilinear, the output seems right, **but  still the losses seems very stable around some
particular values,:****(which I am really confuse about)**

Loss[1] = 0.162387132644659/1260000 ............]  ETA: 23h23s | Step: 65ms     
Loss[2] = 0.86009812355042	
Loss[3] = 1.1286672353745	
Loss[4] = 0.69331330060959	
Loss[5] = 0.69810700416565	
Loss[6] = 0.7666038274765	
Loss[1] = 0.144944995641717/1260000 ............]  ETA: 23h2m | Step: 65ms      
Loss[2] = 0.8347864151001	
Loss[3] = 1.1125816106796	
Loss[4] = 0.6932578086853	
Loss[5] = 0.69813644886017	
Loss[6] = 0.76891779899597	
Loss[1] = 0.169416293501855/1260000 ............]  ETA: 23h3m | Step: 66ms      
Loss[2] = 0.87877112627029	
Loss[3] = 1.1192709207535	
Loss[4] = 0.69345104694366	
Loss[5] = 0.69812732934952	
Loss[6] = 0.76460140943527	
Loss[1] = 0.155231803655623/1260000 ............]  ETA: 23h3m | Step: 66ms      
Loss[2] = 0.83493965864182	
Loss[3] = 1.0719158649445	
Loss[4] = 0.6930724978447	
Loss[5] = 0.6981366276741	
Loss[6] = 0.76579385995865	
Loss[1] = 0.164516046643261/1260000 ............]  ETA: 23h3m | Step: 66ms      
Loss[2] = 0.83939975500107	
Loss[3] = 1.1208488941193	
Loss[4] = 0.69262248277664	
Loss[5] = 0.69816112518311	
Loss[6] = 0.7661435008049	
Loss[1] = 0.150987356901179/1260000 ............]  ETA: 23h4m | Step: 66ms      
Loss[2] = 0.85336416959763	
Loss[3] = 1.1378959417343	
Loss[4] = 0.69268405437469	
Loss[5] = 0.69812244176865	
Loss[6] = 0.76824355125427	
Loss[1] = 0.161790862679487/1260000 ............]  ETA: 23h4m | Step: 66ms      
Loss[2] = 0.86753636598587	
Loss[3] = 1.1050798892975	
Loss[4] = 0.69311159849167	
Loss[5] = 0.69815188646317	
Loss[6] = 0.76528364419937	
Loss[1] = 0.171573653817185/1260000 ............]  ETA: 23h5m | Step: 66ms      
Loss[2] = 0.83807110786438	
Loss[3] = 1.0745347738266	
Loss[4] = 0.69315212965012	
Loss[5] = 0.698102414608	
Loss[6] = 0.76463502645493	
Loss[1] = 0.168636173009873/1260000 ............]  ETA: 23h5m | Step: 66ms      
Loss[2] = 0.89014780521393	
Loss[3] = 1.101013302803	
Loss[4] = 0.69365745782852	
Loss[5] = 0.69804263114929	
Loss[6] = 0.76461642980576	
Loss[1] = 0.156956315040591/1260000 ............]  ETA: 23h5m | Step: 66ms      
Loss[2] = 0.87213122844696	
Loss[3] = 1.0737482309341	
Loss[4] = 0.69354224205017	
Loss[5] = 0.69813603162766	
Loss[6] = 0.76464480161667	
Loss[1] = 0.156559631228459/1260000 ............]  ETA: 23h7m | Step: 66ms      
Loss[2] = 0.82702481746674	
Loss[3] = 1.1217478513718	
Loss[4] = 0.69323301315308	
Loss[5] = 0.69814789295197	
Loss[6] = 0.76717090606689	
Loss[1] = 0.169225648045547/1260000 ............]  ETA: 23h6m | Step: 66ms      
Loss[2] = 0.86106634140015	
Loss[3] = 1.1365866661072	
Loss[4] = 0.69286072254181	
Loss[5] = 0.69812732934952	
Loss[6] = 0.76525324583054	
My four deconv layers: **(note that in the HED network, the deconv learning rate is set to 0)**

       	local side_up2     =  cudnn.SpatialFullConvolution(1, 1, 4, 4, 2, 2, 1, 1):learningRate('weight', 0)
				  :learningRate('bias',0)
				  :weightDecay('weight', 1)
				  :weightDecay('bias', 0)(side_conv2) --(128-1)*2-2*1+4 =256

       local side_up3     =  cudnn.SpatialFullConvolution(1, 1, 8, 8, 4, 4, 2, 2):learningRate('weight', 0)
				  :learningRate('bias',0)
				  :weightDecay('weight', 1)
				  :weightDecay('bias', 0)(side_conv3) --(64-1)*4-2*2+8 =256
        local side_up4     =  cudnn.SpatialFullConvolution(1, 1, 16, 16, 8, 8, 4, 4):learningRate('weight', 0)
				  :learningRate('bias',0)
				  :weightDecay('weight', 1)
				  :weightDecay('bias', 0)(side_conv4)  --(32-1)*8-2*4+16 =256

         local side_up5     =  cudnn.SpatialFullConvolution(1, 1, 32, 32, 16, 16, 8, 8):learningRate('weight', 0)
				  :learningRate('bias',0)
				  :weightDecay('weight', 1)
				  :weightDecay('bias', 0)(side_conv5)  --(16-1)*16-2*8+32 =256

**Comment by nhynes** [comment]
Have you tried freezing the lower layers while you train the deconvs? Here's a quick function that I use to prevent training:
```
function nn.Module:dontTrain(self)
  self.parameters = function() end
  self.accGradParameters = function() end
  return self
end

function nn.Container:dontTrain(self)
  self:applyToModules(function(mod) mod:dontTrain() end)
  return nn.Module.dontTrain(self)
end
```

An alternative is to just zero out the appropriate `gradParameters` before doing the gradient step.

**Comment by brisker** [comment]
@nhynes 
but in the 
[HED](https://github.com/s9xie/hed/blob/master/examples/hed/train_val.prototxt#L147) network, the deconv is not learnable.
I have done the grad:zero() already

**Comment by nhynes** [comment]
Ah, I see. You seem to be using some package that provides the `:learningRate` and `:weightDecay` functions. Perhaps an issue is that the lower layers are trained to use the random weights of the original network's deconvolutions? If you sub in new random weights, I imagine the output will not be what you expect.

The first thing I'd do is to verify that the unmodified pretrained network produces the correct outputs. Only then would I try training (since then I'd know that the bug isn't in the loading step).

**Comment by brisker** [comment]
I tried not to copy weights from vgg16 **(training from scratch),** still the losses looks like this:
**But why? the losses print below are from totally different batches!!**
Loss[2] = 0.7045841217041	
Loss[3] = 0.67549920082092	
Loss[4] = 0.71082574129105	
Loss[5] = 0.70618140697479	
Loss[6] = 0.81400990486145	
Loss[1] = 0.63996106386185/1260000 .............]  ETA: 17h49m | Step: 50ms     
Loss[2] = 0.70423221588135	
Loss[3] = 0.67538571357727	
Loss[4] = 0.71094083786011	
Loss[5] = 0.70626813173294	
Loss[6] = 0.81475788354874	
Loss[1] = 0.63925123214722/1260000 .............]  ETA: 17h45m | Step: 50ms     
Loss[2] = 0.70419818162918	
Loss[3] = 0.67542123794556	
Loss[4] = 0.71089631319046	
Loss[5] = 0.70623272657394	
Loss[6] = 0.81449449062347	
Loss[1] = 0.64067506790161/1260000 .............]  ETA: 17h43m | Step: 50ms     
Loss[2] = 0.70416295528412	
Loss[3] = 0.67565321922302	
Loss[4] = 0.71068668365479	
Loss[5] = 0.70607823133469	
Loss[6] = 0.81310880184174	
Loss[1] = 0.63915413618088/1260000 .............]  ETA: 17h41m | Step: 50ms     
Loss[2] = 0.70463293790817	
Loss[3] = 0.67552381753922	
Loss[4] = 0.71079486608505	
Loss[5] = 0.70615673065186	
Loss[6] = 0.81387656927109	
Loss[1] = 0.64016574621201/1260000 .............]  ETA: 17h40m | Step: 50ms     
Loss[2] = 0.70457464456558	
Loss[3] = 0.67532628774643	
Loss[4] = 0.71099507808685	
Loss[5] = 0.7063068151474	
Loss[6] = 0.81510162353516	
Loss[1] = 0.63892716169357/1260000 .............]  ETA: 17h37m | Step: 50ms     
Loss[2] = 0.70447742938995	
Loss[3] = 0.67550086975098	
Loss[4] = 0.71080732345581	
Loss[5] = 0.70616698265076	
Loss[6] = 0.81396192312241	
Loss[1] = 0.64106488227844/1260000 .............]  ETA: 17h35m | Step: 50ms     
Loss[2] = 0.70477551221848	
Loss[3] = 0.67539966106415	
Loss[4] = 0.71093344688416	
Loss[5] = 0.70626044273376	
Loss[6] = 0.81467467546463

**Comment by brisker** [comment]
@nhynes 
remove all the learning rate parameters for the convolutions still gives the unreasonable losses printed

**Comment by nhynes** [comment]
> (training from scratch),

Then why even load the model using caffegraph? You can just make the full model yourself using the .prototxt.lua and train however you want!

>  losses print below are from totally different batches!

Well, if the batches are all different, I'd expect different losses.

My recommendations are to either 1) use the pretrained weights or 2) train from scratch using not-nngraph because nngraph is slow and over-complicated for what you want.

Also, training VGG from scratch is notoriously difficult--especially when not using batch normalization. If you do go the train-from-scratch route, modify the network to add SpatialBatchNormalization before each ReLU layer (and maybe after each deconv).

**Comment by brisker** [comment]
@nhynes 
But can the model like HED be implemented without nngraph, given its complicated constructure?
6 gradients need to be calculated
https://github.com/s9xie/hed/blob/master/examples/hed/train_val.prototxt

**Comment by nhynes** [comment]
Yes, AFAIK the graph is planar so you should be able to implement it with a clever use of ParallelTable

**Comment by brisker** [comment]
@nhynes 
Hi, it seems that the HED network has modified SigmoidCrossEntropy Loss here:
https://github.com/s9xie/hed/blob/master/src/caffe/layers/sigmoid_cross_entropy_loss_layer.cpp
**and this is clearly not BCECriterion**
I have understood the forward loss computation process by: **(in the handwritten formula ,  p means** **the ground truth and x means the input**)


![image](https://cloud.githubusercontent.com/assets/13804492/24080939/204227ce-0ce4-11e7-830a-47402f408449.png)

![image](https://cloud.githubusercontent.com/assets/13804492/24080929/fcc40d44-0ce3-11e7-8d8a-03573d5a6ebd.png)

but why in this line:https://github.com/s9xie/hed/blob/master/src/caffe/layers/sigmoid_cross_entropy_loss_layer.cpp#L107
the gradients seems like this?

      bottom_diff[i * dim + j] *= 1 * count_neg / (count_pos + count_neg);


 I tried to get the gradients from the handwritten formula above, but I still cannot figure out why the author wrote like that.   Could you please give some advice?
